### PR TITLE
Fix UNC path issue reported in prettier/prettier-vscode#219

### PR DESCRIPTION
**Pull request by azz**
### DO NOT MERGE

I'm experimenting with a way to solve prettier/prettier-vscode#219, where cosmiconfig should stop moving up the directory tree in this case:

```
\\127.0.0.1\c$\a\b\c
\\127.0.0.1\c$\a\b
\\127.0.0.1\c$\a
\\127.0.0.1\c$
STOP - do not read \\127.0.0.1
```

Unfortunately, `path.dirname('\\x\y') === '\\x'`, which is not what we want.

Notes:
1. I wrote this on Mac and will need to run it through CI to see if it passes on Windows.
2. I should test this on an actual Windows machine.
3. Is it ok to have tests that only run on one platform?

### CODE REVIEW

**Comment by sudo-suhas** [comment]
Hey @azz, I work on a windows machine and I can help you out if you need me to test anything. Could you please help me understand the issue and the approach you are trying to fix it?

In my opinion, it's okay to have some tests that only run on windows if it can't be avoided.

Regarding `process.platform === 'win32'`, turns out detecting if the process is running on windows is [not that simple](https://github.com/jonschlinkert/is-windows/blob/63aeae70709733429c026447b7e08039b7ef0ac9/index.js#L16-L22). Could you switch it out to [`is-windows`](https://www.npmjs.com/package/is-windows) instead?

File: test/util.js [hunk 1]
```diff
@@ -49,15 +49,16 @@ exports.mockStatIsDirectory = function mockStatIsDirectory(result) {
...
+  isAbsolute
 ) {
   startCount = startCount || 0;
 
   expect(readFileMock).toHaveBeenCalledTimes(searchPaths.length + startCount);
 
   searchPaths.forEach((searchPath, idx) => {
     expect(readFileMock.mock.calls[idx + startCount][0]).toBe(
-      path.join(__dirname, searchPath)
+      path.join(isAbsolute ? '' : __dirname, searchPath)
```
Comment (sudo-suhas) [inline test/util.js]: How about `isAbsolute ? searchPath : path.join(__dirname, searchPath)`?

**Comment by azz** [comment]
I am actually super confused about this. Node should be smart enough to use `path.win32` as `require('path')`, which **does** handle UNC paths correctly. 

```js
path.win32.dirname('//machine/share$') // -> '//machine/share$'
```

So your code should actually be correct. Minus the cygwin/mingw case...

Perhaps it should be

```patch
-const nextDirectory = path.dirname(directory);
+const nextDirectory = isWindows() ? path.win32.dirname(directory) : path.dirname(directory);
```

**Comment by sudo-suhas** [comment]
Just tried the example you had given:
```
λ node
> process.platform
'win32'
> path.dirname('//machine/share$')
'//machine/share$'
```

What am I missing?

**Comment by azz** [comment]
Yeah, that works exactly as it should. However in prettier-vscode/issues/219, somehow cosmiconfig manages to try to read the equivalent of `//machine`

**Comment by sudo-suhas** [comment]
@azz Does this have anything to do with the wrong kind of slash?

```
> path.dirname('\\127.0.0.1\c$')
'\\'
> path.dirname('//127.0.0.1/c$')
'//127.0.0.1/c$'
```

I don't know enough about network drives to know which would be the right kind of slash 😅.

I [recently changed](https://github.com/davidtheclark/cosmiconfig/pull/91/files#diff-af1dd9c8c993118830d07a83c09d4f2bL138) the logic for getting the next directory. This could fix the issue, though I am not too sure:
```
λ node
> const nextDir = (dir) => {
... const splitPath = dir.split(path.sep);
... return splitPath.length > 1 ? splitPath.slice(0, -1).join(path.sep) : null;
}
undefined
> nextDir('\\127.0.0.1\c$')
''
> path.dirname('\\127.0.0.1\c$')
'\\'
```

**Comment by azz** [comment]
You need four `\`

```js
path.dirname('\\\\127.0.0.1\\c$')
```

**Comment by davidtheclark** [comment]
@sudo-suhas @azz What's the status of this PR?

**Comment by sudo-suhas** [comment]
Not able to replicate the issue(see https://github.com/davidtheclark/cosmiconfig/pull/98#issuecomment-333835175) therefore not sure about merging it either.

**Comment by davidtheclark** [comment]
@sudo-suhas Well, I'd be fine with closing it as stale and waiting to see if the issue arises again. I'll leave it up to you?

**Comment by azz** [comment]
Yeah I ended up out of my depth on this one, found it difficult to replicate and test. It's a fairly minor issue so we can wait to see if it is reported again.

**Comment by GeekMode** [comment]
Found the error. It's actually in the 'resolve' module's nodeModulesPaths function (filename: node-modules-paths.js).

On line 40, they use path.join which removes the initial backslash and resulting in '\\\\IP_ADDRESS\SOME_LOCATION' being turned into '\\IP_ADDRESS\SOME_LOCATION'.

A simple switch from `path.join` to `path.resolve` should correct this issue in all cases I can think of.

**Pull request closed without merging**